libs/ml/examples/cbow_simple/w2v_cbow_dataloader.hpp

```diff
@@ -0,0 +1,247 @@
+    std::map<uint64_t, std::pair<std::string, uint64_t>> reverse_vocab;
+    for (auto const &kvp : vocab_)
+    {
+      reverse_vocab[kvp.second.first] = std::make_pair(kvp.first, kvp.second.second);
+    }
+    for (auto const &sentence : data_)
+    {
+      std::string s;
+      for (auto const &word : sentence)
+      {
+        if (reverse_vocab[word].second >= min)
+        {
+          s += reverse_vocab[word].first + " ";
+        }
+      }
+      new_loader.AddData(s);
+    }
+    data_  = std::move(new_loader.data_);
+    vocab_ = std::move(new_loader.vocab_);
+  }
+
+  void InitUnigramTable()
+  {
+    std::vector<uint64_t> frequencies(VocabSize());
+    for (auto const &kvp : GetVocab())
+    {
+      frequencies[kvp.second.first] = kvp.second.second;
+    }
+    unigram_table_.Reset(1e8, frequencies);
+  }
+
+  ReturnType &GetNext(ReturnType &t)
+  {
+    // This seems to be one of the most important tricks to get word2vec to train
+    // The number of context words changes at each iteration with values in range [1 * 2,
+    // window_size_ * 2]
+    uint64_t dynamic_size = rng_() % window_size_ + 1;
+    t.second.Set(0, 0, T(data_[currentSentence_][currentWord_ + dynamic_size]));
+
+    for (uint64_t i(0); i < dynamic_size; ++i)
+    {
+      t.first.Set(i, 0, T(data_[currentSentence_][currentWord_ + i]));
+      t.first.Set(i + dynamic_size, 0,
+                  T(data_[currentSentence_][currentWord_ + dynamic_size + i + 1]));
+    }
+
+    for (uint64_t i(dynamic_size * 2); i < t.first.size(); ++i)
+    {
+      t.first(i, 0) = -1;
+    }
+
+    for (uint64_t i(1); i < negative_samples_; ++i)
+    {
+      t.second(i, 0) = T(unigram_table_.SampleNegative(static_cast<uint64_t>(t.second(0, 0))));
+    }
+    currentWord_++;
+    if (currentWord_ >= data_.at(currentSentence_).size() - (2 * window_size_))
+    {
+      currentWord_ = 0;
+      currentSentence_++;
+    }
+    return t;
+  }
+
+  ReturnType GetNext() override
+  {
+    fetch::math::Tensor<T> t({window_size_ * 2, 1});
+    fetch::math::Tensor<T> label({negative_samples_, 1});
+    ReturnType             p(t, label);
+    return GetNext(p);
+  }
+
+  std::size_t VocabSize() const
+  {
+    return vocab_.size();
+  }
+
+  bool AddData(std::string const &s)
+  {
+    std::vector<uint64_t> indexes = StringsToIndexes(PreprocessString(s));
+    if (indexes.size() >= 2 * window_size_ + 1)
+    {
+      data_.push_back(std::move(indexes));
+      return true;
+    }
+    return false;
+  }
+
+  std::map<std::string, std::pair<uint64_t, uint64_t>> const &GetVocab() const
+  {
+    return vocab_;
+  }
+
+  std::string WordFromIndex(uint64_t index)
+  {
+    for (auto const &kvp : vocab_)
+    {
+      if (kvp.second.first == index)
+      {
+        return kvp.first;
+      }
+    }
+    return "";
+  }
+
+private:
+  std::vector<uint64_t> StringsToIndexes(std::vector<std::string> const &strings)
+  {
+    std::vector<uint64_t> indexes;
```

Comment:
pedantic, i know, but it's indices (indexes is not english, well, maybe it's a verb but not a noun)